File: src/core/branch/branch.ts
import fs from "node:fs";
import path from "node:path";
import { randomUUID } from "node:crypto";
import { parseDocument } from "yaml";
import { copilotSessionStateDir } from "../paths.js";

/**
 * TS-native session fork, mirroring the session-branch skill's branch_session.py:
 * copy a session directory into a new one, rewrite identity + lineage in
 * workspace.yaml (`branch_of`/`branch_note`), rewrite the session.start event,
 * reset rewind/checkpoint history, and drop the per-session DB and stale locks.
 *
 * IMPORTANT: this only ever CREATES a new directory under the session-state dir;
 * it never modifies the parent (or any existing) session. The copy skips the
 * parent's `session.db*` and lock files so forking a LIVE session is safe.
 */

export interface ForkResult {
  newSessionId: string;
  newSessionName: string;
  newSessionPath: string;
  parentId: string;
}

export interface ForkOptions {
  note?: string;
  /** Override the session-state dir (tests). Defaults to Copilot's. */
  stateDir?: string;
  /** Override the generated id (tests). */
  newId?: string;
}

const CHECKPOINT_INDEX_HEADER =
  "# Checkpoint History\n\n" +
  "Checkpoints are listed in chronological order. Checkpoint 1 is the oldest, " +
  "higher numbers are more recent.\n\n" +
  "| # | Title | File |\n" +
  "|---|-------|------|\n";

const REWIND_INDEX_JSON = '{"version":1,"snapshots":[],"filePathMap":{}}\n';

const SKIP_COPY_RE = /^(session\.db(-shm|-wal)?|inuse\.\d+\.lock)$/;

/** Synchronous sleep (no deps) for the rename retry loop. */
function sleepSync(ms: number): void {
  Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, ms);
}

/**
 * Rename with retry. On Windows a just-written directory can be transiently
 * locked (antivirus, search indexer), making renameSync fail with EPERM/EBUSY;
 * retry a few times before giving up.
 */
function renameWithRetry(from: string, to: string, attempts = 12, delayMs = 50): void {
  for (let i = 0; ; i++) {
    try {
      fs.renameSync(from, to);
      return;
    } catch (err) {
      const code = (err as NodeJS.ErrnoException).code ?? "";
      if (i >= attempts - 1 || !["EPERM", "EBUSY", "ENOTEMPTY", "EACCES"].includes(code)) {
        throw err;
      }
      sleepSync(delayMs);
    }
  }
}

/** Collapse whitespace and truncate to a short single-line title. */
function compact(value: string, maxLen = 72): string {
  const v = (value || "").split(/\s+/).filter(Boolean).join(" ");
  if (v.length <= maxLen) return v;
  return v.slice(0, maxLen - 3).trimEnd() + "...";
}

/** Branch `parentId` into a new session directory. Returns the new identity. */
export function branchSession(parentId: string, opts: ForkOptions = {}): ForkResult {
  const stateDir = opts.stateDir ?? copilotSessionStateDir;
  const currentDir = path.join(stateDir, parentId);
  const workspacePath = path.join(currentDir, "workspace.yaml");
  if (!fs.existsSync(workspacePath)) {
    throw new Error(`Missing workspace metadata: ${workspacePath}`);
  }

  const newId = opts.newId ?? randomUUID();
  const newDir = path.join(stateDir, newId);
  if (fs.existsSync(newDir)) {
    throw new Error(`Branch destination already exists: ${newDir}`);
  }

  const srcDoc = parseDocument(fs.readFileSync(workspacePath, "utf8"));
  const rawName = srcDoc.get("name");
  const rawSummary = srcDoc.get("summary");
  const baseTitle =
    (typeof rawName === "string" && rawName) ||
    (typeof rawSummary === "string" && rawSummary) ||
    `Session ${parentId.slice(0, 8)}`;
  const branchTitle = `Branch: ${compact(String(baseTitle))} [${newId.slice(0, 8)}]`;
  const now = new Date().toISOString();

  fs.mkdirSync(stateDir, { recursive: true });
  const staging = path.join(stateDir, `.tmp-branch-${randomUUID().replace(/-/g, "")}`);

  try {
    // Copy the session, skipping the (possibly locked) DB and lock files.
    fs.cpSync(currentDir, staging, {
      recursive: true,
      filter: (src) => !SKIP_COPY_RE.test(path.basename(src)),
    });

    // Rewrite identity + lineage in workspace.yaml.
    const stagingWorkspace = path.join(staging, "workspace.yaml");
    const doc = parseDocument(fs.readFileSync(stagingWorkspace, "utf8"));
    doc.set("id", newId);
    doc.set("name", branchTitle);
    doc.set("user_named", true);
    doc.set("summary", branchTitle);
    doc.set("created_at", now);
    doc.set("updated_at", now);
    doc.set("branch_of", parentId);
    doc.set(
      "branch_note",
      opts.note && opts.note.trim()
        ? opts.note.trim()
        : `Branched from: ${baseTitle} (${parentId})`,
    );
    fs.writeFileSync(stagingWorkspace, doc.toString(), "utf8");

    rewriteSessionStart(path.join(staging, "events.jsonl"), newId, branchTitle);
    resetRewind(staging);
    resetCheckpoints(staging);

    // Atomic move into place.
    renameWithRetry(staging, newDir);
  } catch (err) {
    fs.rmSync(staging, { recursive: true, force: true });
    throw err;
  }

  return {
    newSessionId: newId,
    newSessionName: branchTitle,
    newSessionPath: newDir,
    parentId,
  };
}

/** Update the session.start event's identity, if events.jsonl exists. */
function rewriteSessionStart(eventsPath: string, newId: string, branchTitle: string): void {
  if (!fs.existsSync(eventsPath)) return;
  const lines = fs.readFileSync(eventsPath, "utf8").split(/\r?\n/).filter((l) => l.trim());
  const out: string[] = [];
  for (const line of lines) {
    let event: Record<string, unknown>;
    try {
      event = JSON.parse(line) as Record<string, unknown>;
    } catch {
      out.push(line);
      continue;
    }
    if (event.type === "session.start") {
      const data = (event.data ?? {}) as Record<string, unknown>;
      data.sessionId = newId;
      if ("alreadyInUse" in data) data.alreadyInUse = false;
      for (const k of ["name", "title", "summary"]) {
        if (k in data) data[k] = branchTitle;
      }
      event.data = data;
    }
    out.push(JSON.stringify(event));
  }
  fs.writeFileSync(eventsPath, out.join("\n") + "\n", "utf8");
}

function resetRewind(staging: string): void {
  const rewindDir = path.join(staging, "rewind-snapshots");
  fs.mkdirSync(rewindDir, { recursive: true });
  fs.writeFileSync(path.join(rewindDir, "index.json"), REWIND_INDEX_JSON, "utf8");
  const backupDir = path.join(rewindDir, "backups");
  if (fs.existsSync(backupDir)) {
    for (const child of fs.readdirSync(backupDir)) {
      fs.rmSync(path.join(backupDir, child), { recursive: true, force: true });
    }
  }
}

function resetCheckpoints(staging: string): void {
  const checkpointDir = path.join(staging, "checkpoints");
  fs.mkdirSync(checkpointDir, { recursive: true });
  fs.writeFileSync(path.join(checkpointDir, "index.md"), CHECKPOINT_INDEX_HEADER, "utf8");
}

File: rust/dcs-core/src/branch.rs
use std::error::Error;
use std::fmt;
use std::fs;
use std::io;
use std::path::{Path, PathBuf};

use serde_json::Value as JsonValue;
use serde_yaml::{Mapping, Value as YamlValue};
use uuid::Uuid;

use crate::model::SessionBranchResult;

const CHECKPOINT_INDEX_HEADER: &str = "# Checkpoint History\n\nCheckpoints are listed in chronological order. Checkpoint 1 is the oldest, higher numbers are more recent.\n\n| # | Title | File |\n|---|-------|------|\n";
const REWIND_INDEX_JSON: &str = "{\"version\":1,\"snapshots\":[],\"filePathMap\":{}}\n";

#[derive(Debug)]
pub enum BranchError {
    Io(io::Error),
    Yaml(serde_yaml::Error),
    Json(serde_json::Error),
    MissingWorkspace(PathBuf),
    DestinationExists(PathBuf),
    InvalidWorkspace,
}

impl fmt::Display for BranchError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Io(error) => write!(formatter, "{error}"),
            Self::Yaml(error) => write!(formatter, "{error}"),
            Self::Json(error) => write!(formatter, "{error}"),
            Self::MissingWorkspace(path) => {
                write!(formatter, "Missing workspace metadata: {}", path.display())
            }
            Self::DestinationExists(path) => write!(
                formatter,
                "Branch destination already exists: {}",
                path.display()
            ),
            Self::InvalidWorkspace => write!(formatter, "Invalid workspace metadata"),
        }
    }
}

impl Error for BranchError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            Self::Io(error) => Some(error),
            Self::Yaml(error) => Some(error),
            Self::Json(error) => Some(error),
            Self::MissingWorkspace(_) | Self::DestinationExists(_) | Self::InvalidWorkspace => None,
        }
    }
}

impl From<io::Error> for BranchError {
    fn from(error: io::Error) -> Self {
        Self::Io(error)
    }
}

impl From<serde_yaml::Error> for BranchError {
    fn from(error: serde_yaml::Error) -> Self {
        Self::Yaml(error)
    }
}

impl From<serde_json::Error> for BranchError {
    fn from(error: serde_json::Error) -> Self {
        Self::Json(error)
    }
}

pub fn branch_session(
    state_dir: impl AsRef<Path>,
    parent_id: &str,
    note: Option<&str>,
    new_id: Option<&str>,
) -> Result<SessionBranchResult, BranchError> {
    let state_dir = state_dir.as_ref();
    let current_dir = state_dir.join(parent_id);
    let workspace_path = current_dir.join("workspace.yaml");
    if !workspace_path.exists() {
        return Err(BranchError::MissingWorkspace(workspace_path));
    }
    let new_id = new_id
        .map(ToOwned::to_owned)
        .unwrap_or_else(|| Uuid::new_v4().to_string());
    let new_dir = state_dir.join(&new_id);
    if new_dir.exists() {
        return Err(BranchError::DestinationExists(new_dir));
    }

    fs::create_dir_all(state_dir)?;
    let staging = state_dir.join(format!(
        ".tmp-branch-{}",
        Uuid::new_v4().to_string().replace('-', "")
    ));
    let result = (|| {
        copy_session_dir(&current_dir, &staging)?;
        let new_name =
            rewrite_workspace(&staging.join("workspace.yaml"), parent_id, &new_id, note)?;
        rewrite_session_start(&staging.join("events.jsonl"), &new_id, &new_name)?;
        reset_rewind(&staging)?;
        reset_checkpoints(&staging)?;
        rename_with_retry(&staging, &new_dir)?;
        Ok(SessionBranchResult {
            new_session_id: new_id,
            new_session_name: new_name,
            new_session_path: new_dir.to_string_lossy().into_owned(),
            parent_id: parent_id.into(),
        })
    })();
    if result.is_err() {
        let _ = fs::remove_dir_all(&staging);
    }
    result
}

fn copy_session_dir(from: &Path, to: &Path) -> Result<(), BranchError> {
    fs::create_dir_all(to)?;
    for entry in fs::read_dir(from)?.flatten() {
        let name = entry.file_name().to_string_lossy().into_owned();
        if should_skip_copy(&name) {
            continue;
        }
        let dest = to.join(entry.file_name());
        if entry.file_type()?.is_dir() {
            copy_session_dir(&entry.path(), &dest)?;
        } else {
            fs::copy(entry.path(), dest)?;
        }
    }
    Ok(())
}

fn should_skip_copy(name: &str) -> bool {
    if name == "session.db" || name == "session.db-shm" || name == "session.db-wal" {
        return true;
    }
    name.starts_with("inuse.") && name.ends_with(".lock")
}

fn rewrite_workspace(
    workspace_path: &Path,
    parent_id: &str,
    new_id: &str,
    note: Option<&str>,
) -> Result<String, BranchError> {
    let raw = fs::read_to_string(workspace_path)?;
    let mut value = serde_yaml::from_str::<YamlValue>(&raw)?;
    let mapping = value
        .as_mapping_mut()
        .ok_or(BranchError::InvalidWorkspace)?;
    let base_title = first_non_empty(mapping, ["name", "summary"])
        .unwrap_or_else(|| format!("Session {}", parent_id.chars().take(8).collect::<String>()));
    let new_name = format!(
        "Branch: {} [{}]",
        compact(&base_title, 72),
        new_id.chars().take(8).collect::<String>()
    );
    let now = now_iso();
    set(mapping, "id", new_id);
    set(mapping, "name", &new_name);
    set_bool(mapping, "user_named", true);
    set(mapping, "summary", &new_name);
    set(mapping, "created_at", &now);
    set(mapping, "updated_at", &now);
    set(mapping, "branch_of", parent_id);
    let branch_note = note
        .filter(|value| !value.trim().is_empty())
        .map(ToOwned::to_owned)
        .unwrap_or_else(|| format!("Branched from: {base_title} ({parent_id})"));
    set(mapping, "branch_note", &branch_note);
    fs::write(workspace_path, serde_yaml::to_string(&value)?)?;
    Ok(new_name)
}

fn rewrite_session_start(path: &Path, new_id: &str, new_name: &str) -> Result<(), BranchError> {
    if !path.exists() {
        return Ok(());
    }
    let raw = fs::read_to_string(path)?;
    let mut out = Vec::new();
    for line in raw.lines().map(str::trim).filter(|line| !line.is_empty()) {
        let mut event = match serde_json::from_str::<JsonValue>(line) {
            Ok(event) => event,
            Err(_) => {
                out.push(line.to_owned());
                continue;
            }
        };
        if event.get("type").and_then(JsonValue::as_str) == Some("session.start") {
            if let Some(data) = event.get_mut("data").and_then(JsonValue::as_object_mut) {
                data.insert("sessionId".into(), JsonValue::String(new_id.into()));
                if data.contains_key("alreadyInUse") {
                    data.insert("alreadyInUse".into(), JsonValue::Bool(false));
                }
                for key in ["name", "title", "summary"] {
                    if data.contains_key(key) {
                        data.insert(key.into(), JsonValue::String(new_name.into()));
                    }
                }
            }
        }
        out.push(serde_json::to_string(&event)?);
    }
    fs::write(path, format!("{}\n", out.join("\n")))?;
    Ok(())
}

fn reset_rewind(staging: &Path) -> Result<(), BranchError> {
    let rewind_dir = staging.join("rewind-snapshots");
    fs::create_dir_all(&rewind_dir)?;
    fs::write(rewind_dir.join("index.json"), REWIND_INDEX_JSON)?;
    let backup_dir = rewind_dir.join("backups");
    if backup_dir.exists() {
        for entry in fs::read_dir(backup_dir)?.flatten() {
            let path = entry.path();
            if entry.file_type()?.is_dir() {
                fs::remove_dir_all(path)?;
            } else {
                fs::remove_file(path)?;
            }
        }
    }
    Ok(())
}

fn reset_checkpoints(staging: &Path) -> Result<(), BranchError> {
    let checkpoint_dir = staging.join("checkpoints");
    fs::create_dir_all(&checkpoint_dir)?;
    fs::write(checkpoint_dir.join("index.md"), CHECKPOINT_INDEX_HEADER)?;
    Ok(())
}

fn rename_with_retry(from: &Path, to: &Path) -> Result<(), BranchError> {
    let mut last_error = None;
    for _ in 0..12 {
        match fs::rename(from, to) {
            Ok(()) => return Ok(()),
            Err(error) => {
                last_error = Some(error);
                std::thread::sleep(std::time::Duration::from_millis(50));
            }
        }
    }
    Err(BranchError::Io(
        last_error.unwrap_or_else(|| io::Error::other("rename failed")),
    ))
}

fn first_non_empty(
    mapping: &Mapping,
    keys: impl IntoIterator<Item = &'static str>,
) -> Option<String> {
    for key in keys {
        if let Some(value) = mapping
            .get(YamlValue::String(key.into()))
            .and_then(YamlValue::as_str)
        {
            if !value.trim().is_empty() {
                return Some(value.to_owned());
            }
        }
    }
    None
}

fn set(mapping: &mut Mapping, key: &str, value: &str) {
    mapping.insert(
        YamlValue::String(key.into()),
        YamlValue::String(value.into()),
    );
}

fn set_bool(mapping: &mut Mapping, key: &str, value: bool) {
    mapping.insert(YamlValue::String(key.into()), YamlValue::Bool(value));
}

fn compact(value: &str, max_len: usize) -> String {
    let compact = value.split_whitespace().collect::<Vec<_>>().join(" ");
    if compact.len() <= max_len {
        compact
    } else {
        format!("{}...", compact[..max_len.saturating_sub(3)].trim_end())
    }
}

fn now_iso() -> String {
    time::OffsetDateTime::now_utc()
        .format(&time::format_description::well_known::Rfc3339)
        .unwrap_or_else(|_| "1970-01-01T00:00:00Z".into())
}
